modules/core/src/test/java/org/apache/ignite/client/ScanQueryCloseTest.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.G;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.mxbean.QueryMXBean;
+import org.apache.ignite.spi.systemview.view.ScanQueryView;
+import org.apache.ignite.spi.systemview.view.SystemView;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.ListeningTestLogger;
+import org.apache.ignite.testframework.LogListener;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+import static org.apache.ignite.internal.managers.systemview.ScanQuerySystemView.SCAN_QRY_SYS_VIEW;
+import static org.apache.ignite.testframework.GridTestUtils.assertThrowsAnyCause;
+
+/**
+ * Tests scan query close.
+ */
+public class ScanQueryCloseTest extends GridCommonAbstractTest {
+    /** Test logger. */
+    private ListeningTestLogger testLog = new ListeningTestLogger(log);
+
+    /** Query filter latch. */
+    private static CountDownLatch filterLatch = new CountDownLatch(1);
+
+    /** @throws Exception If failed. */
+    @Test
+    public void testScanQueryClose() throws Exception {
+        try (Ignite srv1 = Ignition.start(Config.getServerConfiguration().setGridLogger(testLog));
+             Ignite srv2 = Ignition.start(Config.getServerConfiguration().setGridLogger(testLog));
+
+             IgniteClient client = Ignition.startClient(new ClientConfiguration().setAddresses(Config.SERVER))
+        ) {
+            ClientCache<Object, Object> cache = client.getOrCreateCache(DEFAULT_CACHE_NAME);
+
+            for (int i = 0; i < 10; i++)
+                cache.put(i, i);
+
+            ScanQuery<Object, Object> qry = new ScanQuery<>().setFilter((o, o2) -> {
+                try {
+                    filterLatch.countDown();
+
+                    U.sleep(1000);
+                }
+                catch (IgniteInterruptedCheckedException ignored) {
+                    // No-op.
+                }
+
+                return true;
+            });
+
+            LogListener lsnr = LogListener.matches("Iterator has been closed.").build();
+
+            testLog.registerListener(lsnr);
+
+            IgniteInternalFuture<?> fut = GridTestUtils.runAsync(() -> {
+                cache.query(qry).getAll();
+            });
+
+            filterLatch.await(getTestTimeout(), TimeUnit.SECONDS);
+
+            killQueries();
+
+            assertThrowsAnyCause(log, fut::get, IgniteCheckedException.class, null);
+
+            assertTrue(lsnr.check());
+        }
+    }
+
+    /** Kill all queries. */
+    private static void killQueries() {
+        List<Ignite> srvs = G.allGrids();
+
+        assertFalse(srvs.isEmpty());
+
+        QueryMXBean bean = getMxBean(srvs.get(0).name(), "Query",
+            QueryMXBeanImpl.class.getSimpleName(), QueryMXBean.class);
+
+        for (IgniteEx ignite : F.transform(G.allGrids(), ignite -> (IgniteEx)ignite)) {
```

Comment:
Perhaps we can connect to only one server node by thin client to simplify the test (kill all queries only on node we are connected to)